### ISSUE TITLE
[LEAK] Chunked Multipart upload produces memory leaks

### ISSUE THREAD

**Issue by andrey-kozel** [micronaut-projects/micronaut-core#7699]
### Issue description

I have a controller, that recieves a file and a set of fields

<img width="919" alt="image" src="https://user-images.githubusercontent.com/47276289/179497756-bf45b10f-c11a-460d-b923-114dfceb9abc.png">

And I've created a curl command for testing this endpoint
```
curl "http://localhost:8087/api/storage/production%2Fdrive%2F2640%2F1033819%2Fcontent%2Fac4667be031b80e6830b8638becd72bb%2F2e79d1250160d4449cb411a13af4e53" \
      -H 'Accept-Language: ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7' \
      -H 'Connection: keep-alive' \
      -H 'x-spanning-correlationid: d4r32f5s' \
      -H 'Content-Type: multipart/form-data' \
      -H 'Origin: http://localhost:8087' \
      -H 'Transfer-Encoding: chunked' \
      -H 'accept: application/json' \
      -F 'tenantId=1' \
      -F 'resourceId=1' \
      -F 'objectType=MAIL' \
      -F 'encryptionKey=1' \
      -F 'retentionTime=1' \
      -F 'tags={"Compression-Type": "ZSTD"}' \
      -F 'file=@file_beggier_than_100_mb.tar.gz' \
      -v \
      --compressed &
```

I've tried to do it using `micronaut.server.multipart.mixed: true` upload type, since content length is not known, by default it is using Memory file upload and produces Memory Leaks before breakpoint in the controller was hit.

```
LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.\nRecent access records: \nCreated at:
io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:403)
io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:188)
io.netty.buffer.UnsafeByteBufUtil.copy(UnsafeByteBufUtil.java:436)
io.netty.buffer.PooledUnsafeDirectByteBuf.copy(PooledUnsafeDirectByteBuf.java:216)
io.netty.buffer.AbstractByteBuf.copy(AbstractByteBuf.java:1194)
io.netty.handler.codec.http.multipart.HttpPostMultipartRequestDecoder.loadDataMultipartOptimized(HttpPostMultipartRequestDecoder.java:1196)
io.netty.handler.codec.http.multipart.HttpPostMultipartRequestDecoder.getFileUpload(HttpPostMultipartRequestDecoder.java:926)
io.netty.handler.codec.http.multipart.HttpPostMultipartRequestDecoder.decodeMultipart(HttpPostMultipartRequestDecoder.java:572)
io.netty.handler.codec.http.multipart.HttpPostMultipartRequestDecoder.parseBodyMultipart(HttpPostMultipartRequestDecoder.java:463)
io.netty.handler.codec.http.multipart.HttpPostMultipartRequestDecoder.parseBody(HttpPostMultipartRequestDecoder.java:432)
io.netty.handler.codec.http.multipart.HttpPostMultipartRequestDecoder.offer(HttpPostMultipartRequestDecoder.java:347)
io.netty.handler.codec.http.multipart.HttpPostMultipartRequestDecoder.offer(HttpPostMultipartRequestDecoder.java:54)
io.micronaut.http.server.netty.FormDataHttpContentProcessor.onData(FormDataHttpContentProcessor.java:148)
io.micronaut.http.server.netty.AbstractHttpContentProcessor.doOnNext(AbstractHttpContentProcessor.java:79)
io.micronaut.http.server.netty.AbstractHttpContentProcessor.doOnNext(AbstractHttpContentProcessor.java:36)
io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onNext(CompletionAwareSubscriber.java:56)
io.micronaut.http.netty.reactive.HandlerPublisher.publishMessage(HandlerPublisher.java:393)
io.micronaut.http.netty.reactive.HandlerPublisher.flushBuffer(HandlerPublisher.java:470)
io.micronaut.http.netty.reactive.HandlerPublisher.publishMessageLater(HandlerPublisher.java:360)
io.micronaut.http.netty.reactive.HandlerPublisher.channelRead(HandlerPublisher.java:323)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
io.micronaut.http.netty.stream.HttpStreamsHandler.handleReadHttpContent(HttpStreamsHandler.java:316)
io.micronaut.http.netty.stream.HttpStreamsHandler.channelRead(HttpStreamsHandler.java:282)
io.micronaut.http.netty.stream.HttpStreamsServerHandler.channelRead(HttpStreamsServerHandler.java:134)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:99)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
io.netty.handler.codec.http.HttpServerKeepAliveHandler.channelRead(HttpServerKeepAliveHandler.java:64)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
io.netty.handler.flow.FlowControlHandler.dequeue(FlowControlHandler.java:200)
io.netty.handler.flow.FlowControlHandler.channelRead(FlowControlHandler.java:162)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:327)
io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:314)
io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:435)
io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:279)
io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995)
io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
java.base/java.lang.Thread.run(Thread.java:833)
```

I've connected to the java process using visualVM, and I see, that every upload leaves chunk of 4Mb in memory
<img width="1007" alt="image" src="https://user-images.githubusercontent.com/47276289/179498866-942ff20d-f113-4737-bfde-04eddd6cb770.png">

After that I've decided to use `micronaut.server.multipart.disk: true` leak disappeared, but it started to trow OOM, I've tried to debug the issue and I found out, that on NettyHttpRequest.release() it calls buiildBody() and reads whole the file into memory
<img width="1255" alt="image" src="https://user-images.githubusercontent.com/47276289/179499717-eb7a0628-2aaf-495a-99bb-a147730b9da9.png">

All of this chunk wasn't released and after some upload produces out of memory
<img width="985" alt="image" src="https://user-images.githubusercontent.com/47276289/179499994-1c5e95bc-55c3-42f4-9b17-9c5805f6ecf9.png">

OS: `Monterey 12.3.1`

Java: 
```
openjdk 17.0.3 2022-04-19
OpenJDK Runtime Environment Temurin-17.0.3+7 (build 17.0.3+7)
OpenJDK 64-Bit Server VM Temurin-17.0.3+7 (build 17.0.3+7, mixed mode, sharing)

```

Is it a bug, or am I doing something wrong?

**Comment by aglumova** [comment]
Amazing issue, we also had the similar problem.

Still do not know how to resolve it. It would be nice to understand can we improve the Netty wrapper or handle the lazy loading of the file.

Thanks a lot for the help!

**Comment by dstepanov** [comment]
What Micronaut version do you have?

**Comment by andrey-kozel** [comment]
Sorry, I haven't mentioned it.
`<io.micronaut.version>3.5.3</io.micronaut.version>`

**Comment by graemerocher** [comment]
Can you provide an example that reproduces the issue please?

**Comment by andrey-kozel** [comment]
https://github.com/andrey-kozel/micronaut-error-demo Here is an example of the simple controller and tests for it.
In all the test I'm uploading file of ~ 20Mb as bytes array
First test works as expected
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/47276289/179606127-a479d00e-1c4f-4ca4-9292-fb8feb550cf6.png">

The second test is used apache HttpClient MultipartUpload and produces leak in the logs (Happens after a few retries)
<img width="1605" alt="image" src="https://user-images.githubusercontent.com/47276289/179607238-f0c25baf-030f-45ac-bd5b-248c84fca5c8.png">

The third test with works good disk=true throws OOM on 500Xmx 
<img width="1538" alt="image" src="https://user-images.githubusercontent.com/47276289/179606149-2600e179-ecf5-4714-b193-7dac12b7c96b.png">
<img width="347" alt="image" src="https://user-images.githubusercontent.com/47276289/179609284-80f51b6e-7a2e-4920-bba3-106001e5ea04.png">

The fourth example works as expected using apache Http client on the 600Xmx
<img width="529" alt="image" src="https://user-images.githubusercontent.com/47276289/179609129-ee2fd556-1dd5-4994-8bd4-d0da89b58111.png">

curl request also produces leak with file ~20Mb
<img width="1529" alt="image" src="https://user-images.githubusercontent.com/47276289/179609632-5a66208b-bfd3-4441-acb5-73877a5ba2c7.png">
```
 curl "http://localhost:8087/api/storage/production%2Fdrive%2F2640%2F1033819%2Fcontent%2Fac4667be031b80e6830b8638becd72bb%2F2e79d1250160d4449cb411a13af4e53" \
      -H 'Accept-Language: ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7' \
      -H 'Connection: keep-alive' \
      -H 'x-spanning-correlationid: d4r32f5s' \
      -H 'Content-Type: multipart/form-data' \
      -H 'Origin: http://localhost:8087' \
      -H 'Transfer-Encoding: chunked' \
      -H 'accept: application/json' \
      -F 'tenantId=1' \
      -F 'resourceId=1' \
      -F 'objectType=MAIL' \
      -F 'encryptionKey=1' \
      -F 'retentionTime=1' \
      -F 'tags={"Compression-Type": "ZSTD"}' \
      -F 'file=@big-file.tar.gz' \
      -v \
      --compressed
```

Looks like it depends on the client.

**Comment by aglumova** [comment]
Hi @yawkat,

Do you have any ideas about the possible fix and the timeline?

Thanks for your assist here.

**Comment by yawkat** [comment]
i can reproduce the issue, but it might take a bit to find the root cause.

**Comment by aglumova** [comment]
@yawkat Thank you for the feedback!!

**Comment by yawkat** [comment]
Caused by https://github.com/netty/netty/issues/12627

I can fix this in micronaut, but it's ugly, so I'd prefer to wait on a netty fix. Easy workaround is to not use mixed mode for upload management.

**Comment by andrey-kozel** [comment]
@yawkat Thank you! But I've attached a demo where disk upload throws OOM, am I configured client wrong?

**Comment by yawkat** [comment]
there's some issues in the micronaut HTTP client as well but I don't see OOM with the apache client

**Comment by yawkat** [comment]
@sdelamo what's the current timeline for 3.6? I think the netty issue will take a while still, we could implement my workaround for 3.6. it involves copying the relevant classes from netty and adjusting their behavior, so it's a bit hacky, but we could remove them again once the issue is fixed on the netty side.

**Comment by graemerocher** [comment]
@yawkat a temporary workaround implementation is ok

**Comment by andrey-kozel** [comment]
Could you tell us, please, when it is going to be released?

**Comment by yawkat** [comment]
with 3.6

**Comment by andrey-kozel** [comment]
I mean, may be you have some timeline for this release?

**Comment by sdelamo** [comment]
> I mean, may be you have some timeline for this release?

We target by the end of this week.

**Comment by andrey-kozel** [comment]
Thank you!

**Comment by aglumova** [comment]
@sdelamo Could you please share the timeline for this release with the fix?

**Comment by yawkat** [comment]
The plan is today

**Comment by andrey-kozel** [comment]
Was the release postponed for some reasons?

**Comment by yawkat** [comment]
Yes. Please be patient. It's not going to release any faster from asking.

**Comment by andrey-kozel** [comment]
@yawkat @graemerocher I found another issue related to this problem.

I have a controller with definition like this
<img width="891" alt="image" src="https://user-images.githubusercontent.com/47276289/188114144-78f31228-2930-47f9-84ae-cb49ee2f434f.png">

If I'm uploading files and set all the fields it behaves normal, but if I'm not sending field "tags", the next problem happened:
1. AbstractRouteMatch treats it as an unresolved argument
<img width="955" alt="image" src="https://user-images.githubusercontent.com/47276289/188114484-2ba57a47-aa2a-4864-b524-81cce3b97954.png">
2. And again reads body with extremely huge "file" into memory
<img width="823" alt="image" src="https://user-images.githubusercontent.com/47276289/188114725-db4d990d-f433-4050-803a-15cfe360b8fd.png">
3. OOM
<img width="1680" alt="image" src="https://user-images.githubusercontent.com/47276289/188114962-28cf0a95-9479-46bd-a417-fcfdaf7fdce1.png">

**Comment by yawkat** [comment]
This looks unrelated. Please make a new issue

**Comment by andrey-kozel** [comment]
Here is the issue https://github.com/micronaut-projects/micronaut-core/issues/7939
